### DISCUSSION TITLE
connection pooling

### DISCUSSION THREAD

**Discussion by jelmd** [prisma/prisma#20337] [Q&A]
I've read the announcements wrt. 5.x and was baffled about the "need to open new connection", etc. statements. Therefore the question:

Prisma does not use any connection pooling, really?

**Comment by nurul3101** [comment]
Hello @jelmd 👋 

Can you please share which announcements you are referring to?

**Comment by jelmd** [comment]
https://github.com/prisma/prisma/releases => https://www.prisma.io/blog/prisma-5-f66prwkjx72s => https://www.prisma.io/blog/prisma-and-serverless-73hbgKnZ6t

**Reply by janpio** [reply]
Of course, Prisma uses connection pooling. (See also https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/connection-pool)

The blog posts you linked to refer to the cold start scenario, when the pool is still empty and no connections exist yet. (The graphs also include loading the Prisma Client into the application, which of course also is only necessary once.)

**Comment by jelmd** [comment]
Aha, thanx @janpio, good to hear. However, I checked with `ss -tp4` and cannot spot any connection to 127.0.0.1:3306. I use nextjs with
```
DATABASE_URL=mysql://foo:bar@localhost:3306/aapp?connection_limit=10
DATABASE_TMP_URL=mysql:/foo:bar@localhost:3306/aapp_tmp?connection_limit=10
```

Wrt. blog posts: OK, sometimes I head the impression, that in general connection handling is a problem.

BTW: usually one/the OS starts a [web] service and after an expected warmup (which is in most cases I know negligible) it should perform as expected. So I do not understand the hype about fast startup (and related confusion). IMHO it would be much better to make Prisma much smarter wrt. day-by-day use, i.e. optimize queries and related features like switching the protocol, etc. ;-)

**Reply by janpio** [reply]
> However, I checked with ss -tp4 and cannot spot any connection to 127.0.0.1:3306.

Did you open a database connection with `$connect()` or run a query before checking for an open connection? Prisma only opens the first one on demand. But then it keeps it open. If you run multiple queries at the same time, it will also open further connections and keep them in the pool until the pool size limit is hit - then it stops opening new ones and just works with the existing ones.

> BTW: usually one/the OS starts a [web] service and after an expected warmup (which is in most cases I know negligible) it should perform as expected. So I do not understand the hype about fast startup (and related confusion). IMHO it would be much better to make Prisma much smarter wrt. day-by-day use, i.e. optimize queries and related features like switching the protocol, etc. ;-)

In serverless compute environments like serverless functions, every request can run into such a situation. Every time there is no runtime environment available that is not busy, AWS Lambda and Co have to start a new one as each one can only handle _1_ request at a time. So it is not uncommon that hundreds or thousands of serverless functions are spun up in short time - so it is essential that Prisma does not take unnecessary milliseconds in these moments.

**Comment by jelmd** [comment]
Yes. Actually I use a  single `prisma.ts` within a nextjs app, which contains all functions used to CRUD data. It relies on Prisma to open connections as needed. Usage example:
```ts
import { Prisma, PrismaClient } from "@prisma/client";
import type { Adapter } from "next-auth/adapters";
import { PrismaAdapter } from "@next-auth/prisma-adapter";

const globalForPrisma = global as unknown as {
	prisma: PrismaClient | undefined
};

// because PC misbehaves and dumps thrown errors to stdout, we tell it to not log errors. If
// needed, we do it by ourselves.
// @see https://www.prisma.io/docs/guides/other/troubleshooting-orm/help-articles/nextjs-prisma-client-dev-practices
const prisma = globalForPrisma.prisma ?? new PrismaClient({log: [
		'query',	// comment this out to avoid query logging
		'info',
		'warn'
	]});

if (process.env.NODE_ENV !== 'production') {
	globalForPrisma.prisma = prisma;
	/*
	prisma.$on<any>('query', (e: Prisma.QueryEvent) => {
		console.log('Query: ' + e.query)
		console.log('Params: ' + e.params)
		console.log('Duration: ' + e.duration + 'ms')
	});
	*/
}
export function getDbAdapter(): Adapter {
	return PrismaAdapter(prisma);
}

/**
 * Remove all expired entries from the verify table.
 */
export async function cleanupVerify() {
	let ex = new Date();

	try {
		await prisma.verify.deleteMany({
			where: {
				expires: { lte: ex }
			}
		});
	} catch (e) {
		console.log(normalizeException(e).message);
	}
}
// ...
```
Before checking with `ss -tap4` I [re]loaded a page, which triggers DB access, yepp.

BTW: `serverless` sounds weird. Don't know such a IMHO non-sense. I always use a server to serve requests and if something serves a request, doesn't it automagically becomes a **serve**r ... ? ;-)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
